File: examples/assets/neonpulse/audio.lua
-- ****************************************************************
-- NEON PULSE - sound.
--
-- Effects are positioned in the arena rather than played flat, so the
-- side a brick breaks on is the side you hear it: the listener rides the
-- camera, and every trigger passes the world position of whatever caused
-- it. The ambience is the one deliberate exception - a bed that has no
-- location and must not move (see A.startAmbience).
--
-- Every sound is generated, not sampled - see sfx/generate_sfx.py.
-- ****************************************************************

local C = import("config")

local A = { sounds = {}, keep = {} }

local SFX_PATH = GAME_PATH .. "sfx/"

-- Same lifetime rule as everything else in this game: sol owns these
-- objects, so anything the audio engine still points at has to stay
-- reachable from Lua.
local function keep(obj)
	A.keep[#A.keep + 1] = obj
	return obj
end

function A.build()
	G.audio = AudioManager.new()
	keep(G.audio)

	if not G.audio:isInitialized() then
		-- No output device (headless, or one held exclusively elsewhere).
		-- The engine already logged it; the game just runs silent, so this
		-- is not worth a second warning or any special-casing below.
		A.enabled = false
		return
	end

	A.enabled = true
	G.audio:setMasterVolume(C.audio.master)

	local function load(name)
		local voices = C.audio.voices[name] or C.audio.voices.default
		local s = keep(Sound.new(SFX_PATH .. name .. ".wav", voices))
		if s:isLoaded() then
			s:setAttenuation(AttenuationModel.Linear, C.audio.minDistance, C.audio.maxDistance)
		end
		A.sounds[name] = s
	end

	load("paddle"); load("wall"); load("brick"); load("armoured")
	load("powerup"); load("lost"); load("levelclear"); load("launch")

	A.startAmbience()
end

-- The ambience is an AudioSource rather than a Sound: it loops, it needs to
-- persist and be faded, and it must NOT be positioned - a background bed
-- that pans and attenuates as the camera drifts would read as a bug.
-- Spatialization off makes it a constant, locationless layer.
function A.startAmbience()
	local obj = GameObject.new()
	local src = AudioSource.new(SFX_PATH .. "ambience.wav", true)
	obj:addComponent(src)
	G.scene:add(obj)
	keep(obj); keep(src)
	A.ambienceObject, A.ambience = obj, src

	if not src:isLoaded() then return end
	src:setSpatialization(false)
	src:setLooping(true)
	src:setVolume(C.audio.ambienceVolume)
	-- Faded in so the very first frame does not start with an abrupt hum.
	src:fadeIn(1500)
end

-- Builds the ball's spatialized loop. Returned rather than attached here so
-- entities.lua can own it alongside the rest of the ball's components.
function A.newBallHum()
	if not A.enabled then return nil end
	local src = keep(AudioSource.new(SFX_PATH .. "hum.wav", false))
	if not src:isLoaded() then return nil end
	src:setSpatialization(true)
	src:setLooping(true)
	src:setVolume(0)
	src:setAttenuation(AttenuationModel.Linear, C.audio.minDistance, C.audio.maxDistance)
	return src
end

-- Maps ball speed onto the hum's pitch, so a ball that has sped up through a
-- level is audibly faster as well as visually.
function A.setHumSpeed(src, speed01)
	if src == nil then return end
	local a = C.audio.humPitchAtBaseSpeed
	local b = C.audio.humPitchAtMaxSpeed
	src:setPitch(a + (b - a) * speed01)
end

-- Positioned trigger. `x`/`y` are arena coordinates; the play plane's z is
-- filled in here so no caller has to think about it.
function A.playAt(name, x, y, pitch)
	if not A.enabled then return end
	local s = A.sounds[name]
	if s == nil or not s:isLoaded() then return end
	s:playAt(Vec3.new(x, y, C.arena.playZ), C.audio.volume[name] or 1.0, pitch or 1.0)
end

-- Flat trigger, for events that belong to the game rather than to a place
-- in the arena (level cleared, ball lost).
function A.play(name, pitch)
	if not A.enabled then return end
	local s = A.sounds[name]
	if s == nil or not s:isLoaded() then return end
	s:play(C.audio.volume[name] or 1.0, pitch or 1.0)
end

-- Called once per frame, after the scene update, so the listener reads a
-- current world transform rather than last frame's.
function A.update()
	if not A.enabled then return end
	G.audio:setListenerFromGameObject(G.camera)
end

return A

File: examples/web/neonpulse/audio.js
/** Neon Pulse audio — port of assets/neonpulse/audio.lua
 *
 * Browser autoplay policy: do not construct AudioManager / AudioSource until
 * the first user gesture (space / click). ScriptProcessorNode warnings come
 * from miniaudio's WebAudio backend and are harmless.
 */

export function createAudio(P, G, C, sfxPath) {
  const A = {
    sounds: {},
    keep: [],
    enabled: false,
    built: false,
    ambienceStarted: false,
  };

  function keep(obj) {
    A.keep.push(obj);
    return obj;
  }

  /** No-op at boot — real init is ensureReady() after a gesture. */
  function build() {}

  function ensureReady(entities) {
    if (A.built) return A.enabled;
    A.built = true;

    G.audio = keep(new P.AudioManager());
    if (!G.audio.isInitialized()) {
      A.enabled = false;
      return false;
    }
    A.enabled = true;
    G.audio.setMasterVolume(C.audio.master);

    const load = (name) => {
      const voices = C.audio.voices[name] ?? C.audio.voices.default;
      const s = keep(new P.Sound(`${sfxPath}${name}.wav`, voices));
      if (s.isLoaded()) {
        s.setAttenuation(P.AttenuationModel_Linear, C.audio.minDistance, C.audio.maxDistance);
      }
      A.sounds[name] = s;
    };

    ["paddle", "wall", "brick", "armoured", "powerup", "lost", "levelclear", "launch"].forEach(load);

    const wall = A.sounds.wall;
    if (wall && wall.isLoaded()) wall.setDelayFull(0.11, 0.32, 0.45, 1.0);

    if (entities && entities.balls) {
      for (const b of entities.balls) {
        if (!b.hum) {
          const hum = newBallHum();
          if (hum) {
            b.go.addAudioSource(hum);
            b.hum = hum;
          }
        }
      }
    }

    startAmbience();
    return true;
  }

  function startAmbience() {
    if (!A.enabled || A.ambienceStarted) return;
    A.ambienceStarted = true;

    const obj = new P.GameObject();
    const src = new P.AudioSource(`${sfxPath}ambience.wav`, true);
    obj.addAudioSource(src);
    G.scene.add(obj);
    keep(obj);
    keep(src);
    A.ambienceObject = obj;
    A.ambience = src;

    if (!src.isLoaded()) return;
    src.setSpatialization(false);
    src.setLooping(true);
    src.setVolume(C.audio.ambienceVolume);
    src.setEQQ(P.AudioEQType_LowShelf, 220.0, 4.0, 1.0);
    src.fadeIn(1500);
  }

  function newBallHum() {
    if (!A.enabled) return null;
    const src = keep(new P.AudioSource(`${sfxPath}hum.wav`, false));
    if (!src.isLoaded()) return null;
    src.setSpatialization(true);
    src.setLooping(true);
    src.setVolume(0);
    src.setAttenuation(P.AttenuationModel_Linear, C.audio.minDistance, C.audio.maxDistance);
    return src;
  }

  function setHumSpeed(src, speed01) {
    if (!src) return;
    const a = C.audio.humPitchAtBaseSpeed;
    const b = C.audio.humPitchAtMaxSpeed;
    src.setPitch(a + (b - a) * speed01);
  }

  function playAt(name, x, y, pitch = 1.0) {
    if (!A.enabled) return;
    const s = A.sounds[name];
    if (!s || !s.isLoaded()) return;
    s.playAtVolumePitch(
      new P.Vec3(x, y, C.arena.playZ),
      C.audio.volume[name] ?? 1.0,
      pitch
    );
  }

  function play(name, pitch = 1.0) {
    if (!A.enabled) return;
    const s = A.sounds[name];
    if (!s || !s.isLoaded()) return;
    s.playVolumePitch(C.audio.volume[name] ?? 1.0, pitch);
  }

  function update(dt) {
    if (!A.enabled) return;
    G.audio.setListenerFromGameObjectDt(G.camera, dt);
  }

  A.build = build;
  A.ensureReady = ensureReady;
  A.startAmbience = startAmbience;
  A.newBallHum = newBallHum;
  A.setHumSpeed = setHumSpeed;
  A.playAt = playAt;
  A.play = play;
  A.update = update;
  return A;
}
